services/crawler/crawler.go

```diff
@@ -0,0 +1,104 @@
+package crawler
+
+import (
+	"strings"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/beego/beego/v2/core/logs"
+)
+
+const googleSearchBaseUrl = "https://www.google.com/search"
+
+var doc *goquery.Document
+var htmlResponse string
+var selectorMapping = map[string]string{
+	"topAdWordAdvertisersCount":   "#tads .uEierd",
+	"topAdWordAdvertisersUrls":    "#tads .Krnil",
+	"totalAdWordAdvertisersCount": ".Krnil",
+	"resultsCount":                "#rso .yuRUbf",
+	"resultsUrls":                 "#rso .yuRUbf > a",
+	"totalLinksCount":             "a",
+}
+
+type CrawlData struct {
+	TopAdWordAdvertisersCount   int
+	TopAdWordAdvertisersUrls    []string
+	TotalAdWordAdvertisersCount int
+	ResultsCount                int
+	ResultsUrls                 []string
+	TotalLinksCount             int
+	Html                        string
+}
+
+// Call this function with your search key, it will return the necessary crawled data
+// Ex: Crawl("Buy laptop")
+func Crawl(searchString string) (data *CrawlData, err error) {
+	searchUrl, err := BuildSearchUrl(searchString)
+	if err != nil {
+		return &CrawlData{}, err
```

Comment:
Return a pointer allows you to return `nil` then how about return `nil` instead of `&CrawlData{}`. IMO it is clearer when reading the code.